### PR TITLE
Added Vengeance Demon Hunter support

### PR DESCRIPTION
**Pull request by ronaldpereira**

### CODE REVIEW

File: src/Parser/VengeanceDemonHunter/CPM_ABILITIES.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+export const SPELL_CATEGORY = {
+  ROTATIONAL: 'Spell',
+  COOLDOWNS: 'Cooldown',
+};
+
+const CPM_ABILITIES = [
+  {
+    spell: SPELLS.SPIRIT_BOMB_TALENT,
+    isActive: combatant => combatant.lv90Talent === SPELLS.SPIRIT_BOMB_TALENT.id,
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CPM_ABILITIES.js]: Tip: you can use `hasTalent` for checking for talents, it's a little bit easier to use.

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    const results = super.generateResults();
+
+    const fightDuration = this.fightDuration;
+
+    const nonDpsTimePercentage = this.modules.alwaysBeCasting.totalDamagingTimeWasted / fightDuration;
+    const deadTimePercentage = this.modules.alwaysBeCasting.totalTimeWasted / fightDuration;
+
+    console.log(this.modules.alwaysBeCasting.totalTimeWasted+'---'+fightDuration);
+
+    if (nonDpsTimePercentage > 0.01) {
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: This threshold seems off when compared to dead time; I'd expect this to be higher than the dead time threshold.

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      results.addIssue({
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belong to the AlwaysBeCasting module; in the `suggestions(when)` method

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        issue: `Your non DPS time can be improved. Try to cast damaging spells more regularly.`,
+        stat: `${Math.round(nonDpsTimePercentage * 100)}% non DPS time (<30% is recommended)`,
+        icon: 'petbattle_health-down',
+        importance: getIssueImportance(nonDpsTimePercentage, 0.4, 0.45, true),
+      });
+    }
+    if (deadTimePercentage > 0.2) {
+      results.addIssue({
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belong to the AlwaysBeCasting module; in the `suggestions(when)` method

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        issue: `Your dead GCD time can be improved. Try to Always Be Casting (ABC).`,
+        stat: `${Math.round(deadTimePercentage * 100)}% dead GCD time (<20% is recommended)`,
+        icon: 'spell_mage_altertime',
+        importance: getIssueImportance(deadTimePercentage, 0.35, 0.4, true),
+      });
+    }
+    const castEfficiency = getCastEfficiency(CPM_ABILITIES, this);
+    castEfficiency.forEach((cpm) => {
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this should no longer be necessary, see CastEfficiency.js in the Holy Paladin modules for all that's needed. The suggestion code is in the core CastEfficiency class.

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        results.addIssue({
+          issue: <span>Try to cast <SpellLink id={cpm.ability.spell.id} /> more often ({cpm.casts}/{cpm.maxCasts} casts: {Math.round(cpm.castEfficiency * 100)}% cast efficiency). The recommended cast efficiency is {Math.round(cpm.recommendedCastEfficiency * 100)}%. {cpm.ability.extraSuggestion || ''}</span>,
+          icon: cpm.ability.spell.icon,
+          importance: cpm.ability.importance || getIssueImportance(cpm.castEfficiency, cpm.recommendedCastEfficiency - 0.05, cpm.recommendedCastEfficiency - 0.15),
+        });
+      }
+    });
+
+    results.statistics = [
+      <StatisticBox
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belongs to a new module

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        icon={<Icon icon="class_demonhunter" alt="Damage taken" />}
+        value={`${formatNumber(this.totalDamageTaken / this.fightDuration * 1000)} DTPS`}
+        label='Damage taken'
+        tooltip={`Damage taken breakdown:
+            <ul>
+              ${Object.keys(this.damageBySchool).reduce((v, x) => {return v+=`<li>${getMagicDescription(x)} damage taken ${formatThousands(this.damageBySchool[x])} (${formatPercentage(this.damageBySchool[x]/this.totalDamageTaken)}%)</li>`; }, '')}
+            </ul>
+            Total damage taken ${formatThousands(this.totalDamageTaken)}`}
+      />,
+      <StatisticBox
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belongs to a new module

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            src="/img/healing.png"
+            style={{ border: 0 }}
+            alt="Healing"
+          />
+        )}
+        value={`${formatNumber(this.totalHealing / this.fightDuration * 1000)} HPS`}
+        label='Healing done'
+        tooltip={`The total healing done was ${formatThousands(this.totalHealing)}.`}
+      />,
+      <StatisticBox
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belongs to the AlwaysBeCasting module in the `statistic()` function

File: src/Parser/VengeanceDemonHunter/CombatLogParser.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        icon={<Icon icon="spell_mage_altertime" alt="Dead GCD time" />}
+        value={`${formatPercentage(deadTimePercentage)} %`}
+        label='Dead GCD time'
+        tooltip='Dead GCD time is available casting time not used. This can be caused by latency, cast interrupting, not casting anything (e.g. due to movement/stunned), etc.'
+      />,
+      <StatisticBox
```
Comment (MartijnHols) [inline src/Parser/VengeanceDemonHunter/CombatLogParser.js]: With the new module setup I believe this belongs to a new module

**Comment by sMteX** [comment]
If in doubt, look how it's implemented in the Holy Paladin module, that one should be most up-to-date. I've been looking into that a lot for inspiration.

**Comment by ronaldpereira** [comment]
I'll surely adapt this code into the new module setup. Now that I know more about the system, I'm more comfortable to change it as well. Thanks for the suggestions and I'll be updating it asap.

**Comment by MartijnHols** [comment]
@sMteX That's right, and it's a terrible way to work and for that I'm sorry. I really do have plans to make proper documentation instead where this is at least mentioned but preferably the way to work is explained in detail instead, but that keeps on getting delayed in favor of first improving the ways to work so that it needn't be rewritten. Maybe I should just write the outlines asap and iterate and increment on that as I revamp other parts of the code.

**Comment by sMteX** [comment]
@MartijnHols Nah, don't worry about it too much. It is a bit more difficult sometimes, but us developers like a challenge, as long as it's not impossible, right? If you have ideas or improvements, stick to them as long as you have them in your mind. Documentation can wait a bit, and if things are unclear, we can always ask on Discord.

**Pull request merged**